File: src/main/kotlin/com/prescription/service/ClinicalDecisionService.kt
package com.prescription.service

import com.prescription.domain.Prescription
import io.micronaut.core.annotation.Introspected
import io.micronaut.serde.annotation.Serdeable
import jakarta.inject.Singleton
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit

@Introspected
@Serdeable
data class DrugInteraction(
    val severity: InteractionSeverity,
    val description: String,
    val medications: List<String>
)

enum class InteractionSeverity {
    MILD, MODERATE, SEVERE
}

@Introspected
@Serdeable
data class DosageAlert(
    val type: DosageAlertType,
    val message: String,
    val recommendation: String
)

enum class DosageAlertType {
    MAX_DOSE_EXCEEDED, FREQUENCY_TOO_HIGH, DURATION_TOO_LONG
}

@Introspected
@Serdeable
data class ClinicalAlert(
    val prescriptionId: String,
    val patientId: String,
    val alertType: ClinicalAlertType,
    val severity: AlertSeverity,
    val message: String,
    val recommendation: String,
    val createdAt: LocalDateTime
)

enum class ClinicalAlertType {
    DRUG_INTERACTION, DOSAGE_ALERT, ALLERGY_ALERT
}

enum class AlertSeverity {
    LOW, MEDIUM, HIGH, CRITICAL
}

@Singleton
open class ClinicalDecisionService {

    private val drugInteractions = mapOf(
        "warfarin" to listOf(
            DrugInteraction(
                severity = InteractionSeverity.SEVERE,
                description = "Increased bleeding risk when combined with aspirin",
                medications = listOf("aspirin", "acetylsalicylic acid")
            )
        ),
        "amoxicillin" to listOf(
            DrugInteraction(
                severity = InteractionSeverity.MODERATE,
                description = "May reduce effectiveness of oral contraceptives",
                medications = listOf("ethinyl estradiol", "levonorgestrel")
            )
        )
    )

    private val maxDosages = mapOf(
        "acetaminophen" to MaxDosage(4000.0, "mg", "24h"),
        "ibuprofen" to MaxDosage(3200.0, "mg", "24h"),
        "amoxicillin" to MaxDosage(6000.0, "mg", "24h")
    )

    private val maxDurations = mapOf(
        "prednisone" to 14,
        "hydrocodone" to 7,
        "oxycodone" to 7
    )

    data class MaxDosage(val max: Double, val unit: String, val period: String)

    fun checkDrugInteractions(newPrescription: Prescription, existingPrescriptions: List<Prescription>): List<DrugInteraction> {
        val interactions = mutableListOf<DrugInteraction>()
        val newMedication = newPrescription.medicationName.lowercase()
        
        val potentialInteractions = drugInteractions[newMedication] ?: emptyList()
        
        for (interaction in potentialInteractions) {
            for (existingPrescription in existingPrescriptions) {
                if (existingPrescription.status.name == "ACTIVE") {
                    val existingMedication = existingPrescription.medicationName.lowercase()
                    if (interaction.medications.contains(existingMedication)) {
                        interactions.add(interaction)
                    }
                }
            }
        }
        
        return interactions
    }

    fun checkDosageAlerts(prescription: Prescription): List<DosageAlert> {
        val alerts = mutableListOf<DosageAlert>()
        val medication = prescription.medicationName.lowercase()
        val maxDosage = maxDosages[medication]
        
        if (maxDosage != null && prescription.unit == maxDosage.unit) {
            val dailyDose = (24.0 / prescription.frequencyHours) * prescription.dosage
            
            if (dailyDose > maxDosage.max) {
                alerts.add(
                    DosageAlert(
                        type = DosageAlertType.MAX_DOSE_EXCEEDED,
                        message = "Daily dose of ${dailyDose}${prescription.unit} exceeds maximum recommended dose of ${maxDosage.max}${maxDosage.unit}",
                        recommendation = "Consider reducing dose or frequency. Maximum daily dose should not exceed ${maxDosage.max}${maxDosage.unit}."
                    )
                )
            }
        }
        
        if (prescription.frequencyHours < 4) {
            alerts.add(
                DosageAlert(
                    type = DosageAlertType.FREQUENCY_TOO_HIGH,
                    message = "Dosing frequency is very high (less than 4 hours between doses)",
                    recommendation = "Consider extending the interval between doses to reduce risk of side effects."
                )
            )
        }
        
        val maxDuration = maxDurations[medication]
        if (maxDuration != null) {
            val durationDays = ChronoUnit.DAYS.between(prescription.startDate, prescription.endDate)
            
            if (durationDays > maxDuration) {
                alerts.add(
                    DosageAlert(
                        type = DosageAlertType.DURATION_TOO_LONG,
                        message = "Treatment duration of $durationDays days exceeds recommended maximum of $maxDuration days",
                        recommendation = "Consider limiting treatment duration to $maxDuration days or provide additional monitoring."
                    )
                )
            }
        }
        
        return alerts
    }

    fun generateClinicalAlerts(prescription: Prescription, existingPrescriptions: List<Prescription>): List<ClinicalAlert> {
        val alerts = mutableListOf<ClinicalAlert>()
        val now = LocalDateTime.now()
        
        val interactions = checkDrugInteractions(prescription, existingPrescriptions)
        for (interaction in interactions) {
            alerts.add(
                ClinicalAlert(
                    prescriptionId = prescription.prescriptionId,
                    patientId = prescription.patientId,
                    alertType = ClinicalAlertType.DRUG_INTERACTION,
                    severity = mapSeverity(interaction.severity),
                    message = "Drug interaction detected: ${interaction.description}",
                    recommendation = "Review interaction between ${prescription.medicationName} and ${interaction.medications.joinToString(", ")}. Consider alternative medications or additional monitoring.",
                    createdAt = now
                )
            )
        }
        
        val dosageAlerts = checkDosageAlerts(prescription)
        for (dosageAlert in dosageAlerts) {
            alerts.add(
                ClinicalAlert(
                    prescriptionId = prescription.prescriptionId,
                    patientId = prescription.patientId,
                    alertType = ClinicalAlertType.DOSAGE_ALERT,
                    severity = mapDosageAlertSeverity(dosageAlert.type),
                    message = dosageAlert.message,
                    recommendation = dosageAlert.recommendation,
                    createdAt = now
                )
            )
        }
        
        return alerts
    }

    private fun mapSeverity(interactionSeverity: InteractionSeverity): AlertSeverity {
        return when (interactionSeverity) {
            InteractionSeverity.MILD -> AlertSeverity.LOW
            InteractionSeverity.MODERATE -> AlertSeverity.MEDIUM
            InteractionSeverity.SEVERE -> AlertSeverity.CRITICAL
        }
    }

    private fun mapDosageAlertSeverity(alertType: DosageAlertType): AlertSeverity {
        return when (alertType) {
            DosageAlertType.MAX_DOSE_EXCEEDED -> AlertSeverity.CRITICAL
            DosageAlertType.FREQUENCY_TOO_HIGH -> AlertSeverity.HIGH
            DosageAlertType.DURATION_TOO_LONG -> AlertSeverity.MEDIUM
        }
    }
}

File: src/services/ClinicalDecisionService.ts
import { Prescription } from '../types';

export interface DrugInteraction {
  severity: 'MILD' | 'MODERATE' | 'SEVERE';
  description: string;
  medications: string[];
}

export interface DosageAlert {
  type: 'MAX_DOSE_EXCEEDED' | 'FREQUENCY_TOO_HIGH' | 'DURATION_TOO_LONG';
  message: string;
  recommendation: string;
}

export interface ClinicalAlert {
  prescriptionId: string;
  patientId: string;
  alertType: 'DRUG_INTERACTION' | 'DOSAGE_ALERT' | 'ALLERGY_ALERT';
  severity: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  message: string;
  recommendation: string;
  createdAt: string;
}

export class ClinicalDecisionService {
  private drugInteractions: Map<string, DrugInteraction[]>;
  private maxDosages: Map<string, { max: number; unit: string; period: string }>;
  private maxDurations: Map<string, number>;

  constructor() {
    this.drugInteractions = new Map();
    this.maxDosages = new Map();
    this.maxDurations = new Map();
    this.initializeClinicalData();
  }

  private initializeClinicalData(): void {
    this.drugInteractions.set('warfarin', [
      {
        severity: 'SEVERE',
        description: 'Increased bleeding risk when combined with aspirin',
        medications: ['aspirin', 'acetylsalicylic acid']
      }
    ]);

    this.drugInteractions.set('amoxicillin', [
      {
        severity: 'MODERATE',
        description: 'May reduce effectiveness of oral contraceptives',
        medications: ['ethinyl estradiol', 'levonorgestrel']
      }
    ]);

    this.maxDosages.set('acetaminophen', { max: 4000, unit: 'mg', period: '24h' });
    this.maxDosages.set('ibuprofen', { max: 3200, unit: 'mg', period: '24h' });
    this.maxDosages.set('amoxicillin', { max: 6000, unit: 'mg', period: '24h' });

    this.maxDurations.set('prednisone', 14);
    this.maxDurations.set('hydrocodone', 7);
    this.maxDurations.set('oxycodone', 7);
  }

  async checkDrugInteractions(newPrescription: Prescription, existingPrescriptions: Prescription[]): Promise<DrugInteraction[]> {
    const interactions: DrugInteraction[] = [];
    const newMedication = newPrescription.medicationName.toLowerCase();
    
    const potentialInteractions = this.drugInteractions.get(newMedication) || [];
    
    for (const interaction of potentialInteractions) {
      for (const existingPrescription of existingPrescriptions) {
        if (existingPrescription.status === 'ACTIVE') {
          const existingMedication = existingPrescription.medicationName.toLowerCase();
          if (interaction.medications.includes(existingMedication)) {
            interactions.push(interaction);
          }
        }
      }
    }
    
    return interactions;
  }

  async checkDosageAlerts(prescription: Prescription): Promise<DosageAlert[]> {
    const alerts: DosageAlert[] = [];
    const medication = prescription.medicationName.toLowerCase();
    const maxDosage = this.maxDosages.get(medication);
    
    if (maxDosage && prescription.unit === maxDosage.unit) {
      const dailyDose = (24 / prescription.frequencyHours) * prescription.dosage;
      
      if (dailyDose > maxDosage.max) {
        alerts.push({
          type: 'MAX_DOSE_EXCEEDED',
          message: `Daily dose of ${dailyDose}${prescription.unit} exceeds maximum recommended dose of ${maxDosage.max}${maxDosage.unit}`,
          recommendation: `Consider reducing dose or frequency. Maximum daily dose should not exceed ${maxDosage.max}${maxDosage.unit}.`
        });
      }
    }
    
    if (prescription.frequencyHours < 4) {
      alerts.push({
        type: 'FREQUENCY_TOO_HIGH',
        message: 'Dosing frequency is very high (less than 4 hours between doses)',
        recommendation: 'Consider extending the interval between doses to reduce risk of side effects.'
      });
    }
    
    const maxDuration = this.maxDurations.get(medication);
    if (maxDuration) {
      const startDate = new Date(prescription.startDate);
      const endDate = new Date(prescription.endDate);
      const durationDays = Math.ceil((endDate.getTime() - startDate.getTime()) / (1000 * 60 * 60 * 24));
      
      if (durationDays > maxDuration) {
        alerts.push({
          type: 'DURATION_TOO_LONG',
          message: `Treatment duration of ${durationDays} days exceeds recommended maximum of ${maxDuration} days`,
          recommendation: `Consider limiting treatment duration to ${maxDuration} days or provide additional monitoring.`
        });
      }
    }
    
    return alerts;
  }

  async generateClinicalAlerts(prescription: Prescription, existingPrescriptions: Prescription[]): Promise<ClinicalAlert[]> {
    const alerts: ClinicalAlert[] = [];
    const now = new Date().toISOString();
    
    const interactions = await this.checkDrugInteractions(prescription, existingPrescriptions);
    for (const interaction of interactions) {
      alerts.push({
        prescriptionId: prescription.prescriptionId,
        patientId: prescription.patientId,
        alertType: 'DRUG_INTERACTION',
        severity: this.mapSeverity(interaction.severity),
        message: `Drug interaction detected: ${interaction.description}`,
        recommendation: `Review interaction between ${prescription.medicationName} and ${interaction.medications.join(', ')}. Consider alternative medications or additional monitoring.`,
        createdAt: now
      });
    }
    
    const dosageAlerts = await this.checkDosageAlerts(prescription);
    for (const dosageAlert of dosageAlerts) {
      alerts.push({
        prescriptionId: prescription.prescriptionId,
        patientId: prescription.patientId,
        alertType: 'DOSAGE_ALERT',
        severity: this.mapDosageAlertSeverity(dosageAlert.type),
        message: dosageAlert.message,
        recommendation: dosageAlert.recommendation,
        createdAt: now
      });
    }
    
    return alerts;
  }

  private mapSeverity(interactionSeverity: 'MILD' | 'MODERATE' | 'SEVERE'): 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL' {
    switch (interactionSeverity) {
      case 'MILD': return 'LOW';
      case 'MODERATE': return 'MEDIUM';
      case 'SEVERE': return 'CRITICAL';
      default: return 'MEDIUM';
    }
  }

  private mapDosageAlertSeverity(alertType: 'MAX_DOSE_EXCEEDED' | 'FREQUENCY_TOO_HIGH' | 'DURATION_TOO_LONG'): 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL' {
    switch (alertType) {
      case 'MAX_DOSE_EXCEEDED': return 'CRITICAL';
      case 'FREQUENCY_TOO_HIGH': return 'HIGH';
      case 'DURATION_TOO_LONG': return 'MEDIUM';
      default: return 'MEDIUM';
    }
  }
}